Address book send

chikeichan opened the pull request:
- [x] Implement new send to-autocomplete UI
- [x] Implement new send flow - address list view
- [x] Implement new send flow - send content view
- [x] Fix confirm flow if necessary
- [x] Implement "Add to Address" book modal
- [x] Implement Error/Loading state for to-autocomplete
- [x] Add/Fix unit tests
- [x] Add/Fix integration tests

chikeichan committed:
Style Send Header

chikeichan committed:
Move Send to-row to send view and restyle

chikeichan committed:
Add "Recents" group to select recipient view

chikeichan committed:
Rename SendToRow to AddRecipient

chikeichan committed:
Basic UI and Layout

chikeichan committed:
New ENSInput component

chikeichan committed:
wip - fuzzy search for input

chikeichan committed:
small refactor

chikeichan commented:
![addressend](https://user-images.githubusercontent.com/8507735/59013920-1eca8380-87f0-11e9-9230-876491d5ecd6.gif)

chikeichan committed:
Add Dialog

chikeichan committed:
Click to open modal

chikeichan committed:
Create AddToAddressBookModal component

chikeichan committed:
Modal styling and layout

chikeichan committed:
modal i18n

chikeichan committed:
Add to Addressbook

chikeichan committed:
ens wip

chikeichan committed:
ens wip

chikeichan committed:
ENS Resolution

chikeichan committed:
Reset input

chikeichan committed:
Send to explicit address

chikeichan committed:
Happy Path Complete

chikeichan commented:
https://www.loom.com/share/8715b92315db4712847416f719997e65
https://www.loom.com/share/519c3b244ae14796ba2d7045c49e611f

chikeichan committed:
Add back error checking

chikeichan committed:
Reset send-to when emptying input

chikeichan committed:
Add back warning object

chikeichan committed:
Fix linter

chikeichan committed:
Fix unit test #1 - fix import paths

chikeichan committed:
Remove dead tests

chikeichan committed:
One more to go

chikeichan committed:
Fix all unit tests

chikeichan committed:
add unit test for reducers and actions

chikeichan committed:
test rendering AddRecipient

chikeichan committed:
Add tests for dialog boxes in AddRecipient

chikeichan committed:
Add test for validating

chikeichan committed:
Fix linter

chikeichan committed:
Fix e2e tests

bdresser commented:
@chikeichan this looks awesome!! could you record a short demo of the flow in the extension popup when you get a chance? will help with design QA

chikeichan committed:
Token send e2e fix

chikeichan commented:
> @chikeichan this looks awesome!! could you record a short demo of the flow in the extension popup when you get a chance? will help with design QA

https://www.loom.com/share/cd3cae924c3748f5ba490eda2de3e801

There you go!

For this one, please still pull it down and play with it locally if possible 🙇 . This is a major refactor of the `send` flow, so we want to be very cautious.

chikeichan committed:
Fix e2e

chikeichan committed:
Fix from-import-beta-ui e2e spec

cjeria commented:
Nice work @chikeichan! Here's a list of some feedback I caught while testing.

1. When clicking “send” from the account view, focus cursor in the to field (if possible) for faster recipient inputting.

2. Make section header say "add recipient” by default. Change to “Send” when ETH address detected.

2.1 Update placeholder text to the suggested text in designs

![image](https://user-images.githubusercontent.com/597770/60219865-66fa2600-9831-11e9-8fcc-f89b34f4ca0b.png)

3. Make recipient name 14 pt font size
![image](https://user-images.githubusercontent.com/597770/60219925-9741c480-9831-11e9-96e6-40f7a77c563e.png)

4. Behavior when pasting an ethereum address, don’t make me have to click the list item below the to field. Auto show the send view.
![image](https://user-images.githubusercontent.com/597770/60219964-b2accf80-9831-11e9-90cb-8a4ee3544502.png)

5. If no other tokens in my wallet, don’t make this field a dropdown
![image](https://user-images.githubusercontent.com/597770/60220013-dbcd6000-9831-11e9-807e-49ef10328819.png)

6. In order to avoid this overlap text, we agreed that it would be fine to ellipsis the middle part of an address like "0x28b…6581"
![image](https://user-images.githubusercontent.com/597770/60220019-e25bd780-9831-11e9-8f65-e313803b8c95.png)


7. In full-screen mode, let’s fix the grey bar color so it blends with the rest of the background
![image](https://user-images.githubusercontent.com/597770/60220024-e7b92200-9831-11e9-9ef9-f0aa28c9e76c.png)

8. Are we able to show the saved contact’s name in the confirm screen (if contact was saved)?
![image](https://user-images.githubusercontent.com/597770/60220035-edaf0300-9831-11e9-9957-6b951c65201b.png)

9. Give each row here a top padding of 16px
![image](https://user-images.githubusercontent.com/597770/60220348-f6540900-9832-11e9-8f2b-f093e350b2dd.png)

10. some addresses overlap the "X" here, not sure why? I believe Roboto is mono-spaced... 
![image](https://user-images.githubusercontent.com/597770/60220389-27343e00-9833-11e9-9ed8-849701169594.png)

11. Saving is not working consistently, I was only able to save one contact the first time only. 

I realize this might be an early stage version of the feature, so please let me know what's still being worked on. cc @bdresser

danjm commented:
@chikeichan I think it would be appropriate to remove the `test-integration-flat-chrome` send test at this point, as long as create an issue that documents the behaviour covered by those tests that is not covered by e2e tests. We need to replace these tests with additions to the e2e tests eventually, and this is a significant enough change that I think now is the appropriate time.

danjm commented on ui/app/pages/send/send-content/add-recipient/ens-input.js:
```diff
@@ -0,0 +1,274 @@
+  }
+}
+
+export default connect(
```
In most places we follow the pattern of creating a `.container.js` file for connection to redux. Do you have a preference for doing this within the same file that defines the component?

chikeichan replied:
Fixed

danjm commented:
I think the code looks good generally. I need to take one more close look at the ens-input changes.

danfinlay commented:
This is awesome! It looks great.

A question:
- Is there a way to copy the address after the ENS name has resolved?

Some notes for future improvement, I could open as design issues, would that be right, @cjeria?

- Confirmation screen does not show the account nickname like the send form did.
- The nickname should appear somewhere in the TX history list, when we have a nickname for who we sent to.

cjeria commented:
> * Is there a way to copy the address after the ENS name has resolved?
Which screen/step are you wanting to copy the address?

> Some notes for future improvement, I could open as design issues, would that be right, @cjeria?

Sure! Let's create a new issue in design for future enhancements to this feature.

danfinlay commented:
Opened new issues #105, #106, maybe should've been one issue, feel free to consolidate.

bdresser commented:
@chikeichan this looks awesome! super smooth

Biggest thing I'm noticing is that addresses should be network-specific. If I save a contact on Mainnet, it's not necessarily a testnet address, so showing it in the testnet Send flow could cause a lot of user errors. Let me know if you want to chat about this live or if we should pull @jennypollack in as well.


Some smaller notes:

* I think the "copy" request is in scope for this. On the confirm screen, the user should be able to copy the address their sending to (whether raw address, contact, or ENS name) by hovering over the address field. We can reuse the existing `Copy Address` component (see below). Many folks are religious about double double checking.
<img width="203" alt="Screen Shot 2019-06-28 at 11 33 18 AM" src="https://user-images.githubusercontent.com/1016190/60363479-8d86a100-9998-11e9-89d9-84876169cf34.png">



* On the "Add Contact" screen (below):
<img width="363" alt="Screen Shot 2019-06-28 at 11 28 16 AM" src="https://user-images.githubusercontent.com/1016190/60363414-4bf5f600-9998-11e9-9c20-d8c2a1a3e9ca.png">

  * pressing `enter` on the "add contact" screen should save the contact
  * we should auto-focus cursor inside the text field when this modal opens

Also agree with Christian's notes above. Happy to set something up for next week if you want to go over any of this live!

chikeichan committed:
Make section header say "add recipient” by default

chikeichan committed:
Auto-focus add recipient input

chikeichan committed:
Update placeholder text

chikeichan committed:
Update input title font size

chikeichan committed:
Auto advance to next step if user paste a valid address

chikeichan committed:
Ellipsify address when recipient is selected

chikeichan committed:
Fix app header background color on desktop

chikeichan committed:
Give each form row a margin of 16px

chikeichan committed:
Use .container/.component naming pattern for ens-input

chikeichan committed:
Auto-focus on input when add to addressbook modal is opened; Save on Enter

chikeichan commented:
@cjeria 
1. Fixed
2. Fixed
3. Fixed
4. Fixed
5. Created issue #6787 
6. Fixed
7. Fixed
8. Created issue #6788 
9. Fixed
10. Roboto is not monospaced. Roboto Mono is monospaced.
11. Tagging @jennypollack here for attention re: saving inconsistency.

@danjm 
- Refactor the naming to follow `.container.js/.component.js`
- Would love some help on integration test if possible. Been trying to make it pass in CI and locally...

@bdresser 
- "clicked to save" is already there on confirm screen
- Added autofocus for `AddAddressModal`
- Added "Press Enter to Save" for `AddAddressModal`

chikeichan committed:
Fix and add unit test

chikeichan committed:
Fix selectors name in e2e tests

danjm committed:
Correct e2e test token amount for address-book-send changes

danjm committed:
Adds e2e test for editing a transaction

danjm committed:
Delete test/integration/lib/send-new-ui.js

danjm committed:
Add tests for amount max button and high value error on send screen to test/e2e/metamask-ui.spec.js

danjm commented:
- [ ] Proper ENS error message when not on correct network

danjm committed:
Improve ENS message when not found on current network

danjm committed:
Add error to indicate when network does not support ENS

danjm commented:
@cjeria Improved error messaging related to ENS:

![enserrors](https://user-images.githubusercontent.com/7499938/61762592-15e83e00-adad-11e9-8904-ec3e41bf149b.gif)

whymarrh commented:
Closed by #6914
